merge/list_test.go

```diff
@@ -0,0 +1,282 @@
+	if err != nil {
+		t.Fatalf("Failed to create parser: %v", err)
+	}
+
+	t.Run("disjointItemsOrder", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- d
+- e
+`)
+
+		output := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+- e
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("overlappingItemsOrder", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- a
+- b
+- d
+`)
+
+		output := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("disjointOverlappingItemsOrder", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- b
+- a
+- d
+`)
+
+		output := typed.YAMLObject(`
+items:
+- b
+- a
+- c
+- d
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("orderViolationHandling1", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- d
+- b
+- a
+`)
+
+		output := typed.YAMLObject(`
+items:
+- d
+- b
+- a
+- c
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("orderViolationHandling2", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+- e
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- e
+- d
+- b
+- a
+`)
+
+		output := typed.YAMLObject(`
+items:
+- c
+- e
+- d
+- b
+- a
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("orderViolationHandling3", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+- d
+- e
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- a
+- d
+- b
+- e
+`)
+
+		output := typed.YAMLObject(`
+items:
+- a
+- c
+- d
+- b
+- e
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+
+	t.Run("orderConstraintCheck", func(t *testing.T) {
+		state := &State{
+			Updater:  &merge.Updater{},
+			Parser:   parser,
+			Typename: "lists",
+		}
+
+		lhs := typed.YAMLObject(`
+items:
+- a
+- b
+- c
+`)
+
+		rhs := typed.YAMLObject(`
+items:
+- a
+- x
+- y
+- z
+- b
+`)
+
+		output := typed.YAMLObject(`
+items:
+- a
+- x
+- y
+- z
+- b
+- c
+`)
+
+		twoWayMerge(t, state, lhs, rhs, output)
+	})
+}
+
+func twoWayMerge(t *testing.T, state *State, lhs, rhs, output typed.YAMLObject) {
```

Comment:
I think the way to interpret those tests is that lhs == initial condition and rhs == a thing that is applied. Which is effectively two applies in a row, exactly like you've written here. But the names are confusing in this context. So maybe:

s/twoWayMerge/singleApplyTest/
s/lhs/initialObject/
s/rhs/appliedObject/

Reply:
I guess that would work yes.